crates/nu-parser/src/type_check.rs

```diff
@@ -9,22 +9,19 @@ use nu_protocol::{
 
 pub fn type_compatible(lhs: &Type, rhs: &Type) -> bool {
     // Structural subtyping
-    let is_compatible = |expected: &[(String, Type)], found: &[(String, Type)]| {
-        if expected.is_empty() || found.is_empty() {
-            // We treat an incoming empty table/record type as compatible for typechecking purposes
-            // It is the responsibility of the runtime to reject if necessary
-            true
-        } else if expected.len() > found.len() {
-            false
-        } else {
-            expected.iter().all(|(col_x, ty_x)| {
-                if let Some((_, ty_y)) = found.iter().find(|(col_y, _)| col_x == col_y) {
-                    type_compatible(ty_x, ty_y)
-                } else {
-                    false
-                }
-            })
-        }
+    let is_compatible = |expected: &[(String, Type)], found: &[(String, Type)]| -> bool {
+        // We treat an incoming empty table/record type as compatible for typechecking purposes
+        // It is the responsibility of the runtime to reject if necessary
+        found.is_empty()
+            || expected.is_empty()
+            || (expected.len() <= found.len()
+                && expected.iter().all(|(col_x, ty_x)| {
+                    found
+                        .iter()
+                        .find(|(col_y, _)| col_y == col_x)
+                        .filter(|(_, ty_y)| type_compatible(ty_x, ty_y))
+                        .is_some()
```

Comment:
If I grok the new logic correctly we could use the recently stabilized [`Option::is_some_and`](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_and) as I first had to wrap my head around the `find`/`filter` chain.

Reply:
yeah, `is_some_and()` is more coinsise, i didn't know that they got it into stable. let me use it